Handle IConstructorBodyOperation and IMethodBodyOperation in CFG

mavasani opened the pull request:
Also added corresponding overloads to SemanticModel.GetControlFlowGraph

333fred commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -1221,6 +1225,68 @@ private void VisitStatements(IEnumerable<IOperation> statements)
+                VisitStatement(operation.Initializer);
+            }
+
+            if (haveLocals)
```
This is not correct. Locals declared via declaration patterns in a constructor initializer _are_ visible in the body.

333fred replied:
For example, this is completely valid code (under C# 7.3):
```C#
class C : Base
{
    C() : base(out var i)
    {
        O(i);
    }
    void O(int i)
    {
        i = 1;
    }
}

class Base
{
    protected Base(out int i)
    {
        i = 1;
    }
}
```

---
In reply to: [189345445](https://github.com/dotnet/roslyn/pull/26955#discussion_r189345445) [](ancestors = 189345445)

mavasani replied:
Ah, I got confused by the doc comment: http://source.roslyn.io/#Microsoft.CodeAnalysis/Operations/IConstructorBodyOperation%2520.cs,21. Let me adjust the implementation and test baselines.

333fred replied:
Please also add a test like the one above, that way cfg verifier will fail if in the future we break this.

---
In reply to: [189346487](https://github.com/dotnet/roslyn/pull/26955#discussion_r189346487) [](ancestors = 189346487)

mavasani replied:
@333fred I can add a test, but note that existing test `ConstructorBody_14` does guard this case. The fact that the declared local is used or not in the body is not relevant.

333fred replied:
Yes it is. @AlekseyTs ensured that the CFG Verifier will verify that all locals used are contained in the surrounding scope, and it would have failed the above test case.

---
In reply to: [189379223](https://github.com/dotnet/roslyn/pull/26955#discussion_r189379223) [](ancestors = 189379223)

jaredpar commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -1221,6 +1225,69 @@ private void VisitStatements(IEnumerable<IOperation> statements)
+            return null;
+        }
+
+        private void VisitMethodBodyBaseOperation(IMethodBodyBaseOperation operation)
```
Consider as a local function.

jaredpar commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs:
```diff
@@ -281,5 +432,606 @@ class C
+    Statements (0)
+");
+        }
     }
```
Consider a test where the base call has control flow:

``` 
public C(int? i) : base (i ?? 42)
{

}
```

mavasani replied:
I was in two minds about whether such a test will be useful, as the actual rewrite doesn't touch or special case operations within the initializer or body/expression blocks. I think it is worthwhile to at least have one test with flow in initializer, I will add one.

jaredpar replied:
I think this could be meaningfully interesting to have a single case here. The transition between ctor, base, and body has historically been a troublesome area for the compiler. I've introduced bugs there at least twice 😄 

Also in this particular case the two statements are not linked as an array. It's linking two separate nodes on an `IOperation` tree. Seems meaningful to ensure we get the right hookup here.

jaredpar commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs:
```diff
@@ -281,5 +432,606 @@ class C
+            string source = @"
+class C
+{
+    public C() : base()
```
Is constructor chaining `: this()` going to be done in a separate PR?

mavasani replied:
That shouldn't affect the flow graph, its just a different invocation in the initializer. I will still cover it in the new test I plan to add for your above suggestion.

jaredpar replied:
Agree it shouldn't be meaningfully different. Still would like to have at least one test to ensure we exercise the code path.

mavasani commented:
@dotnet-bot retest windows_debug_unit32_prtest please

mavasani commented:
@dotnet-bot retest windows_release_unit32_prtest please

mavasani commented:
@dotnet-bot retest windows_debug_vs-integration_prtest please

mavasani commented:
@dotnet-bot retest windows_release_vs-integration_prtest please

AlekseyTs commented on src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs:
```diff
@@ -1045,11 +1045,15 @@ protected static void VerifyFlowGraphForTest<TSyntaxNode>(CSharpCompilation comp
+        {
+            var model = compilation.GetSemanticModel(syntaxNode.SyntaxTree);
+            ControlFlowGraph graph = ControlFlowGraphVerifier.GetControlFlowGraph(syntaxNode, model);
+            ControlFlowGraphVerifier.VerifyGraph(model.Compilation, expectedFlowGraph, graph);
```
>model.Compilation [](start = 49, length = 17)

```compilation```?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -1221,6 +1225,69 @@ private void VisitStatements(IEnumerable<IOperation> statements)
+                if (operation.ExpressionBody != null)
+                {
+                    // Link last block of visited BlockBody to the exit block.
+                    BasicBlock lastBlock = _blocks.Last();
```
>BasicBlock lastBlock = _blocks.Last(); [](start = 20, length = 38)

We always link CurrentBasicBlock

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs:
```diff
@@ -150,8 +196,39 @@ public C() : base()
+Block[B2] - Block [UnReachable]
+    Predecessors (0)
+    Statements (1)
+        IExpressionStatementOperation (OperationKind.ExpressionStatement, Type: null, IsImplicit) (Syntax: 'throw null')
```
>IExpressionStatementOperation (OperationKind.ExpressionStatement, Type: null, IsImplicit) (Syntax: 'throw null') [](start = 8, length = 112)

Where does this come from?

AlekseyTs replied:
I see what is going on. It looks like we need to adjust implementation of VisitExpressionStatement to return null if its child is a Throw operation, after visiting the child of course. Could you please make this small adjustment, since this affects baselines that you are adding?

---
In reply to: [189704520](https://github.com/dotnet/roslyn/pull/26955#discussion_r189704520) [](ancestors = 189704520)

mavasani replied:
I was going to come talk to you about the difference between expression bodied and block case, thanks for suggesting the correct approach.

---
In reply to: [189708626](https://github.com/dotnet/roslyn/pull/26955#discussion_r189708626) [](ancestors = 189708626,189704520)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IConstructorBodyOperation.cs:
```diff
@@ -214,8 +307,32 @@ public C()
+Block[B2] - Block [UnReachable]
+    Predecessors (0)
+    Statements (1)
+        IExpressionStatementOperation (OperationKind.ExpressionStatement, Type: null, IsImplicit) (Syntax: 'throw null')
```
>IExpressionStatementOperation (OperationKind.ExpressionStatement, Type: null, IsImplicit) (Syntax: 'throw null') [](start = 8, length = 112)

Same issue, it looks like this is already handled in the previous basic block

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IMethodBodyOperation.cs:
```diff
@@ -128,6 +170,431 @@ public void M()
+    Block[B2] - Block [UnReachable]
+        Predecessors (0)
+        Statements (1)
+            IExpressionStatementOperation (OperationKind.ExpressionStatement, Type: null, IsInvalid, IsImplicit) (Syntax: 'j = i')
```
>IExpressionStatementOperation (OperationKind.ExpressionStatement, Type: null, IsInvalid, IsImplicit) (Syntax: 'j = i') [](start = 12, length = 118)

I cannot tell two assignments apart, please use distinct syntax

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IMethodBodyOperation.cs:
```diff
@@ -128,6 +170,431 @@ public void M()
+            var tree = compilation.SyntaxTrees.Single();
+            var node1 = tree.GetRoot().DescendantNodes().OfType<BaseMethodDeclarationSyntax>().Single();
+
+            compilation.VerifyOperationTree(node1, expectedOperationTree:
```
>compilation.VerifyOperationTree(node1, expectedOperationTree: [](start = 12, length = 61)

Same suggestion to get rid of tree validation in new Control Flow tests.

AlekseyTs commented:
Done with review pass (iteration 5)

mavasani commented:
@AlekseyTs addressed feedback

---
In reply to: [390789537](https://github.com/dotnet/roslyn/pull/26955#issuecomment-390789537) [](ancestors = 390789537)
